docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,235 @@
+
+To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `typelib`. A `typelib` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, the `typelib` library will closely mimic the implementation of static types in Luau Analysis. Most importantly, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 
+
+<details><summary>typelib library methods (dropdown)</summary>
+
+Methods under a different type heading (ex: `Singleton`) imply that the methods are only available for those types. At the implementation level, there is a check to make sure that the type-specific methods are being called on the correct types. For instance, `getindexer()` asserts that `istable()` is true.
+
+#### typelib
+All attributes of newly created typelib are initialized with empty tables / arrays and `typelib:getnil()`. For instance, `typelib:newtable()` initializes its properties with an empty table and index / index result type as `typelib:getnil()`.
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getnil()` | `typelib` | returns an immutable runtime representation of the built-in type `nil` |
+| `getunknown()` | `typelib` | returns an immutable runtime representation of the built-in type `unknown` |
+| `getnever()` | `typelib` | returns an immutable runtime representation of the built-in type `never` |
+| `getany()` | `typelib` | returns an immutable runtime representation of the built-in type `any` |
+| `getnegation(arg: typelib)` | `typelib` | returns an immutable runtime representation of the negation of the argument; the argument cannot be `istable()`, `ismetatable` or `isfunction()` |
+| `getboolean()` | `typelib` | returns an immutable runtime representation of the built-in type `boolean` |
+| `getnumber()` | `typelib` | returns an immutable runtime representation of the built-in type `number` |
+| `getstring()` | `typelib` | returns an immutable runtime representation of the built-in type `string` |
+| `getstringsingleton(arg: string)` | `typelib` | returns an immutable runtime representation of a string singleton type of the argument |
+| `getbooleansingleton(arg: boolean)` | `typelib` | returns an immutable runtime representation of a boolean singleton type of the argument |
+| `getunion(arg: {typelib})` | `typelib` | returns an immutable runtime representation of union type of its argument |
+| `getintersection(arg: {typelib})` | `typelib` | returns an immutable runtime representation of intersection type of its argument |
+| `newtable()` | `typelib` | returns a mutable runtime representation of a `table` type |
+| `newmetatable()` | `typelib` | returns a mutable runtime representation of `Metatable` |
```

Comment:
I think I'd like for metatables to be represented as a special blessed property of tables, @andyfriesen has described to me the belief that choosing otherwise (i.e. to make metatable its own sort of type) was a mistake.